Python/importdl.c

```diff
@@ -472,6 +472,19 @@ _PyImport_RunModInitFunc(PyModInitFunction p0,
                     &res, _Py_ext_module_loader_result_ERR_MISSING_DEF);
             goto error;
         }
+
+#ifdef Py_GIL_DISABLED
+        if (res.def->m_size == -1) {
+            // This module may be recreated (without running its init function)
+            // in reload_singlephase_extension(), so remember its GIL slot
+            // here.
+            if (_PyImport_SetModuleGIL(m, ((PyModuleObject *)m)->md_gil) < 0) {
+                _Py_ext_module_loader_result_set_error(
+                        &res, _Py_ext_module_loader_result_EXCEPTION);
+                goto error;
+            }
+        }
+#endif
```

Comment:
I'd also expect this in `import_run_extension()` in import.c, rather than here.

Reply:
This ended up becoming part of the extension cache data, based on another comment, so it's spread between `update_global_state_for_extension()` and `_extensions_cache_set()` rather than the line I linked in my previous comment.